homeassistant/components/yamaha_musiccast/media_player.py

```diff
@@ -1,291 +1,371 @@
+from .const import DEFAULT_ZONE, DOMAIN, REPEAT_MODE_MAPPING
 
-SUPPORTED_FEATURES = (
-    SUPPORT_PLAY
-    | SUPPORT_PAUSE
-    | SUPPORT_STOP
-    | SUPPORT_PREVIOUS_TRACK
-    | SUPPORT_NEXT_TRACK
-    | SUPPORT_TURN_ON
-    | SUPPORT_TURN_OFF
+PARALLEL_UPDATES = 1
+
+
+async def async_setup_entry(
+    hass: HomeAssistantType,
+    entry: ConfigEntry,
+    async_add_entities: Callable[[list[Entity], bool], None],
+) -> None:
+    """Set up MusicCast sensor based on a config entry."""
+    coordinator: MusicCastDataUpdateCoordinator = hass.data[DOMAIN][entry.entry_id]
+
+    name = coordinator.data.network_name
+
+    media_players: list[Entity] = []
+
+    for zone in coordinator.data.zones:
+        zone_name = name if zone == DEFAULT_ZONE else f"{name} {zone}"
+
+        media_players.append(
+            MusicCastMediaPlayer(zone, zone_name, entry.entry_id, coordinator)
+        )
+
+    async_add_entities(media_players, True)
+
+
+MUSIC_PLAYER_SUPPORT = (
+    SUPPORT_PAUSE
     | SUPPORT_VOLUME_SET
     | SUPPORT_VOLUME_MUTE
+    | SUPPORT_TURN_ON
+    | SUPPORT_TURN_OFF
+    | SUPPORT_CLEAR_PLAYLIST
+    | SUPPORT_PLAY
+    | SUPPORT_SHUFFLE_SET
+    | SUPPORT_REPEAT_SET
+    | SUPPORT_PREVIOUS_TRACK
+    | SUPPORT_NEXT_TRACK
+    | SUPPORT_SELECT_SOUND_MODE
     | SUPPORT_SELECT_SOURCE
+    | SUPPORT_STOP
 )
 
-KNOWN_HOSTS_KEY = "data_yamaha_musiccast"
-INTERVAL_SECONDS = "interval_seconds"
 
-DEFAULT_PORT = 5005
-DEFAULT_INTERVAL = 480
+class MusicCastMediaPlayer(MediaPlayerEntity, MusicCastDeviceEntity):
+    """A demo media players."""
 
-PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
-    {
-        vol.Required(CONF_HOST): cv.string,
-        vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
-        vol.Optional(INTERVAL_SECONDS, default=DEFAULT_INTERVAL): cv.positive_int,
-    }
-)
+    def __init__(self, zone_id, name, entry_id, coordinator, device_class=None):
+        """Initialize the demo device."""
+        self._name = name
+        self._player_state = STATE_PLAYING
+        self._volume_muted = False
+        self._shuffle = False
+        self._device_class = device_class
+        self._zone_id = zone_id
+        self.coordinator: MusicCastDataUpdateCoordinator = coordinator
+
+        self._volume_min = self.coordinator.data.zones[self._zone_id].min_volume
+        self._volume_max = self.coordinator.data.zones[self._zone_id].max_volume
+
+        self._cur_track = 0
+        self._repeat = REPEAT_MODE_OFF
+
+        super().__init__(
+            entry_id=entry_id,
+            coordinator=coordinator,
+            name=name,
+            icon="mdi:speaker",
+        )
+
+        self.coordinator.entities.append(self)
+
+    async def async_added_to_hass(self):
+        """Run when this Entity has been added to HA."""
+        # Sensors should also register callbacks to HA when their state changes
+        self.coordinator.musiccast.register_callback(self.async_write_ha_state)
```

Comment:
Why is there a separate callback besides the coordinator update callback? When overwriting here, we need to call the `async_added_to_hass` in the parent class if we want the coordinator entity to work.

Reply:
That is because normally the MusicCast device pushes updates via UDP (local push) but some events never get pushed, thats why we also have the coordinator fetching everything else every 60 seconds